kit/hydra/relay.py

```diff
@@ -0,0 +1,305 @@
+from abc import abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, is_dataclass
+from functools import lru_cache
+import importlib
+import logging
+from pathlib import Path
+import re
+import sys
+from types import ModuleType
+from typing import (
+    Any,
+    ClassVar,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import hydra
+from hydra.utils import instantiate
+from omegaconf import OmegaConf
+from typing_extensions import final
+
+from .utils import SchemaRegistration
+
+__all__ = [
+    "Relay",
+    "Option",
+]
+
+
+@lru_cache()
+def _camel_to_snake(name: str) -> str:
+    name = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", name)
+    return re.sub("([a-z0-9])([A-Z])", r"\1_\2", name).lower()
+
+
+@dataclass(init=False)
+class Option:
+    def __init__(self, class_: Type[Any], name: Optional[str] = None) -> None:
+        self.class_ = class_
+        self._name = name
+
+    @property
+    def name(self) -> str:
+        if self._name is None:
+            cls_name = self.class_.__name__
+            if cls_name.endswith("Conf"):
+                cls_name.rstrip("Conf")
+            return _camel_to_snake(cls_name)
+        return self._name
+
+    @name.setter
+    def name(self, name: Optional[str]) -> None:  # type: ignore
+        self._name = name
+
+
+@dataclass
+class _SchemaImportInfo:
+    schema_name: str
+    name: str
+
+
+R = TypeVar("R", bound="Relay")
+
+
+class Relay:
+    """
+    Abstract class for orchestrating hydra runs.
+
+    This class does away with the hassle of needing to define config-stores, initialise
+    config directories, and manually run configen on classes to convert them into valid schemas.
+    Regular non-hydra compatible, classes can be passed to the `with_hydra` method and
+    configen will be run on them automatically (if use_cached_confs=False or a cached version
+    of the schemas can't be found), with the resulting schemas cached in the config directory.
+
+    Subclasses must implement a 'run' method and will themselves be converted into the
+    primary config classes used to initialise and validate hydra. With the launching of hydra,
+    the subclasses will be instantiated and the run method called on the raw config.
+
+    :example:
+
+    >>>
+    Relay.with_hydra(
```

Comment:
Doesn't this need to be on the same line?

Reply:
No, the block continues until the next empty line